services/vrf/vrf.go

```diff
@@ -0,0 +1,352 @@
+const KeySizeBits = 2048
+const keySizeBytes = KeySizeBits / 8
+const keySizeWords = KeySizeBits / 256
+
+const wordBytes = 32
+
+// PublicExponent is the exponent used in RSA public keys
+// Any change to this forces change to VRF.sol and VRF_test.js, too.
+//
+// A public exponent of 3 is very cheap to calculate, in terms of ethereum gas,
+// and no risk for this application because the seed is uniformly randomly
+// sampled from the ring.
+const PublicExponent = 3
+
+// panicUnless is a basic golang "assert"
+func panicUnless(prop bool, message string) {
+	if !prop {
+		panic(message)
+	}
+}
+
+// Textbook RSA "decryption", copied from crypto/rsa.go/decypt function. Returns
+// seed raised to the private exponent of k, k's modulus. Uses faster CRT method
+// if enabled on k.
+func decrypt(k *rsa.PrivateKey, seed *big.Int) *big.Int {
+	panicUnless(len(k.Primes) == 2,
+		"RSA VRF only works with two-factor moduli.")
+	if k.Precomputed.Dp == nil { // Do it the slow way
+		return new(big.Int).Exp(seed, k.D, k.N)
+	}
+	// We have the precalculated values needed for the CRT.
+	m := new(big.Int).Exp(seed, k.Precomputed.Dp, k.Primes[0])
+	m2 := new(big.Int).Exp(seed, k.Precomputed.Dq, k.Primes[1])
+	m.Sub(m, m2)
+	if m.Sign() < 0 {
+		m.Add(m, k.Primes[0])
+	}
+	m.Mul(m, k.Precomputed.Qinv)
+	m.Mod(m, k.Primes[0])
+	m.Mul(m, k.Primes[1])
+	m.Add(m, m2)
+	return m
+}
+
+// encrypt "encrypts" m under the publicKey, with textbook RSA.
+//
+// This is inadequate for actual encryption. Use rsa.EncryptOAEP for that. The
+// point here is not to encrypt sensitive data, but to use an operation which
+// can only be reversed with knowledge of the private key.
+func encrypt(publicKey *rsa.PublicKey, m *big.Int) *big.Int {
+	exponentBig := new(big.Int).SetUint64(uint64(publicKey.E))
+	return new(big.Int).Exp(m, exponentBig, publicKey.N)
+}
+
+// Proof represents a proof that Output was generated as mandated by the Key
+// from the given Seed, via Decrypt.
+type Proof struct {
+	Key                   *rsa.PublicKey
+	Seed, Decrypt, Output *big.Int
+	bitSize               uint32
+}
+
+// asUint256 returns i represented as an array of packed uint256's, a la solidity
+func asUint256Array(i *big.Int) []byte {
+	inputBytes := i.Bytes()
+	outputBytesDeficit := wordBytes - (len(inputBytes) % wordBytes)
+	if outputBytesDeficit == wordBytes {
+		outputBytesDeficit = 0
+	}
+	rv := append(make([]byte, outputBytesDeficit), inputBytes...)
+	panicUnless(len(rv)%32 == 0 && i.Cmp(new(big.Int).SetBytes(rv)) == 0,
+		"rv is not i as big-endian uint256 array")
+	return rv
+}
+
+func asKeySizeUint256Array(i *big.Int) []byte {
+	a := asUint256Array(i)
+	rv := append(make([]byte, (keySizeBytes-len(a))), a...)
+	panicUnless(len(rv) == keySizeBytes, "i must fit in keySizeBytes")
+	panicUnless(len(rv)%wordBytes == 0, "must generate packed uint256 array")
+	panicUnless(i.Cmp(new(big.Int).SetBytes(rv)) == 0, "must represent i")
```

Comment:
Preferably return an error instead of panicking. Recovering from a panic gets pretty messy, much more manageable to return an error.